airflow/executors/mesos_executor.py

```diff
@@ -0,0 +1,422 @@
+        self.mesos_docker_volume_dag_name = conf.get('mesos', 'DOCKER_VOLUME_DAG_NAME').replace('"', '')
+        self.mesos_docker_volume_dag_container_path = conf.get(
+            'mesos', 'DOCKER_VOLUME_DAG_CONTAINER_PATH'
+        ).replace('"', '')
+        self.mesos_docker_volume_logs_name = conf.get('mesos', 'DOCKER_VOLUME_LOGS_NAME').replace('"', '')
+        self.mesos_docker_volume_logs_container_path = conf.get(
+            'mesos', 'DOCKER_VOLUME_LOGS_CONTAINER_PATH'
+        ).replace('"', '')
+        self.mesos_docker_sock = conf.get('mesos', 'DOCKER_SOCK')
+        self.core_sql_alchemy_conn = conf.get('core', 'SQL_ALCHEMY_CONN')
+        self.core_fernet_key = conf.get('core', 'FERNET_KEY')
+
+    def resource_offers(self, offers):
+        """If we got a offer, run a queued task"""
+        self.log.debug('MESOS OFFER')
+        for i, offer in enumerate(offers):
+            if i == 0:
+                self.run_job(offer)
+            offer.decline()
+            i += 1
+
+    def run_job(self, mesos_offer):
+        """Start a queued Airflow task in Mesos"""
+        offer = mesos_offer.get_offer()
+        tasks = []
+        option = {}
+        offer_cpus = 0
+        offer_mem = 0
+        for resource in offer['resources']:
+            if resource['name'] == "cpus":
+                offer_cpus += resource['scalar']['value']
+            elif resource['name'] == "mem":
+                offer_mem += resource['scalar']['value']
+
+        self.log.debug(
+            "Received offer %s with cpus: %s and mem: %s", offer['id']['value'], offer_cpus, offer_mem
+        )
+
+        remaining_cpus = offer_cpus
+        remaining_mem = offer_mem
+
+        while (
+            (not self.task_queue.empty())
+            and remaining_cpus >= self.task_cpu
+            and remaining_mem >= self.task_mem
+        ):
+
+            key, cmd, executor_config = self.task_queue.get()
+            self.log.debug(executor_config)
+            tid = self.task_counter
+            self.task_counter += 1
+            self.task_key_map[str(tid)] = key
+
+            port_begin = 31000 + tid
+            port_end = 31000 + tid
+
+            self.log.debug("Launching task %d using offer %s", tid, offer['id']['value'])
+
+            task = {
+                'name': "AirflowTask %d" % tid,
+                'task_id': {'value': str(tid)},
+                'agent_id': {'value': offer['agent_id']['value']},
+                'resources': [
+                    {'name': 'cpus', 'type': 'SCALAR', 'scalar': {'value': self.task_cpu}},
+                    {'name': 'mem', 'type': 'SCALAR', 'scalar': {'value': self.task_mem}},
+                    {
+                        'name': 'ports',
+                        'type': 'RANGES',
+                        'ranges': {'range': [{'begin': port_begin, 'end': port_end}]},
+                    },
+                ],
+                'command': {
+                    'shell': 'true',
```

Comment:
Using shell seems like a security risk -- we have to worry about escaping then. Is there another option?

Reply:
Without shell, It was not working with my airflow test container. But, I can understand the security issue. Would it be ok, if I make it configurable? As default "false" and if someone need it, they can change it to true?

Reply:
I make this configurable. The user can choose between true or false.